modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryUtils.cs

```diff
@@ -642,14 +650,175 @@ public static unsafe void WriteString(string val, IBinaryStream stream)
 
             fixed (char* chars = val)
             {
-                int byteCnt = Utf8.GetByteCount(chars, charCnt);
+                int byteCnt = GetUtf8ByteCount(chars, charCnt);
 
                 stream.WriteInt(byteCnt);
 
                 stream.WriteString(chars, charCnt, byteCnt, Utf8);
             }
         }
 
+        /// <summary>
+        /// Converts string to UTF8 bytes.
+        /// </summary>
+        /// <param name="chars">Chars.</param>
+        /// <param name="charCnt">Chars count.</param>
+        /// <param name="byteCnt">Bytes count.</param>
+        /// <param name="enc">Encoding.</param>
+        /// <param name="data">Data.</param>
+        /// <returns>Amount of bytes written.</returns>
+        public static unsafe int StringToUtf8Bytes(char* chars, int charCnt, int byteCnt, Encoding enc, byte* data)
+        {
+            if (!UseStringSerializationVer2)
+                return enc.GetBytes(chars, charCnt, data, byteCnt);
+
+            int strLen = charCnt;
+            int c, cnt;
+
+            int position = 0;
+
+            for (cnt = 0; cnt < strLen; cnt++)
+            {
+                c = *(chars + cnt);
+
+                if (c <= 0x007F)
+                    *(data + position++) = (byte)c;
+                else if (c > 0x07FF)
+                {
+                    *(data + position++) = (byte)(0xE0 | c >> 12 & 0x0F);
+                    *(data + position++) = (byte)(0x80 | c >> 6 & 0x3F);
+                    *(data + position++) = (byte)(0x80 | c & 0x3F);
+                }
+                else
+                {
+                    *(data + position++) = (byte)(0xC0 | c >> 6 & 0x1F);
+                    *(data + position++) = (byte)(0x80 | c & 0x3F);
+                }
+            }
+
+            return position;
+        }
+
+        /// <summary>
+        /// Gets the UTF8 byte count.
+        /// </summary>
+        /// <param name="chars">The chars.</param>
+        /// <param name="strLen">Length of the string.</param>
+        /// <returns>UTF byte count.</returns>
+        private static unsafe int GetUtf8ByteCount(char* chars, int strLen)
+        {
+            int utfLen = 0;
+            int cnt;
+            for (cnt = 0; cnt < strLen; cnt++)
+            {
+                var c = *(chars + cnt);
+
+                // ASCII
+                if (c <= 0x007F)
```

Comment:
Pavel,

I changed this condition on Java side from (c <= 0x007F) to if (c >= 0x0001 && c <= 0x007F). Please update on .Net side.